### ISSUE TITLE
Ports model

### ISSUE THREAD

**Issue by jungkees** [mkruisselbrink/navigator-connect#30]
We've had a couple different proposals for the model to maintain persistent ports.

@mkruisselbrink's original suggestion is to introduce **stashing** ports. As [in the example](https://github.com/mkruisselbrink/navigator-connect/blob/gh-pages/docs/stashed-ports.md#sample-code):

``` js
navigator.connect('https://example.com/services/push')
  .then(function(port) {
      port.postMessage({register: 'apikey'});
      self.stashPort('pushService', port);
    });
```

So, in this model, a client may stash ports so they can get them back when it wants to do a postMessage(). And to my understanding, it seems the ports which are not explicitly stashed will be GC'ed (and so the connection will be severed) when the client and/or the service are unloaded/terminated.

Another perspective that I've thought of is to maintain the connected service as a set of persistent channels that the UA maintains ([See the model here](https://mkruisselbrink.github.io/navigator-connect/#h-model)). With this model, a client does not explicitly stash connected ports but they are inserted into the set by the UA, and the client and the service can only get the port by [`self.ports.match()`](https://mkruisselbrink.github.io/navigator-connect/#h-connectedportcollection). And for this model, the connection is severed when the connected ports are disentangled, which is detected by one of the explicit `port.close()`, unload a Window/Worker client, and unregister a SW client/service.

Would love to have some comments which model we'd want to go with.

**Comment by mkruisselbrink** [comment]
I think a big benefit of something like the stashed ports idea is that it can be used for any `MessagePort`. It doesn't somehow special treat `MessagePort`s that are representing a navigator.connect based connection. What that means is that it is possible to transfer a `MessagePort` from one service worker to another, and still have it persistent; or you could even have a webpage connect to a navigator.connect service, and then pass the `MessagePort` of to its service worker if it wants to persist it.
With some automated bookkeeping it is unclear if/how this will work. It's especially unclear to me how a service worker will be able to identify a particular connection without some explicit service worker provided piece of data associated with the port. And without an explicit step to persist a port (where this explicit step could just be a specific property passed in a property bag to `connect` or `acceptConnection`) I don't see how a service worker can provide such a piece of data.

I think without actually supporting arbitrary MessagePort we lose a lot of the power/reasons to use a MessagePort for this connection in the first place. Also see [this document](https://github.com/mkruisselbrink/navigator-connect/blob/gh-pages/docs/api-idea.md#persisted-messageports) for some more thoughts on how this explicit port persisting API could look like (as well as more thoughts on problems with that approach).

Also I'm rather confused about the actual API you describe to implement the model you're proposing. `ConnectedPortCollection` somehow is exposed in all workers and windows? But how does it even make sense to expose this to anything but service workers? For one it really doesn't make sense for one `MessagePort` to exist in multiple contexts at the same time, so all these workers and windows will need to have separate collections, but then why does anything other than a service worker even need a ConnectedPortCollection?

**Comment by jungkees** [comment]
Just to answer those questions:

> It's especially unclear to me how a service worker will be able to identify a particular connection without some explicit service worker provided piece of data associated with the port.

It's the connection's client's _id_ which identifies the specific connection. (See the note under http://mkruisselbrink.github.io/navigator-connect/#dfn-name-to-connection-map.)

> ConnectedPortCollection somehow is exposed in all workers and windows? But how does it even make sense to expose this to anything but service workers?

Yeah, as you pointed, exposing it to windows and other workers doesn't make sense as the connection will be lost when they are unloaded.

**Comment by jungkees** [comment]
Indeed, now I have better understanding how flexible the port-stashing can work as you explained:

> you could even have a webpage connect to a navigator.connect service, and then pass the MessagePort of to its service worker if it wants to persist it
> I think without actually supporting arbitrary MessagePort we lose a lot of the power/reasons to use a MessagePort for this connection in the first place.

Let's continue to work on your proposal!

**Comment by mkruisselbrink** [comment]
So does something like the API described in [docs/api-idea.md](https://github.com/mkruisselbrink/navigator-connect/blob/gh-pages/docs/api-idea.md) seem like the right way to go then?

**Comment by kinu** [comment]
Relevant blink bug: https://crbug.com/483108
Looks like currently chrome/blink's implementing docs/api-idea.md model

**Comment by kinu** [comment]
Any updates on spec status for this yet?

**Comment by jungkees** [comment]
We're planning to update the spec next week after having some feedback on docs/api-idea.md in a scheduled meeting.
